src/Datastore/RAMDataStore.cpp

```diff
@@ -0,0 +1,337 @@
+  if (it_admin != mAdminQueue.tracks.end()) {
+    pAdminTrack = &(*it_admin);
+  }
+  auto it_normal =
+      find(mNormalQueue.tracks.begin(), mNormalQueue.tracks.end(), track);
+  if (it_normal != mNormalQueue.tracks.end()) {
+    pNormalTrack = &(*it_normal);
+  }
+
+  // User found, look for Track in vote vector
+  auto it_track = find(it->votes.begin(), it->votes.end(), tID);
+  if (it_track != it->votes.end()) {
+    // Track already found in vote vector
+    if (vote) {
+      // track already in vote vector and we want to upvote it: this is a
+      // duplicate, do nothing
+    } else {
+      // Track already in vote vector and we want to remove the upvote:
+      // we want to remove it from upvoted tracks, so remove it from vector of
+      // upvoted tracks and update vote counter in track
+      it->votes.erase(it_track);
+      // decrement its upvote counter
+      if (pAdminTrack != nullptr) {
+        pAdminTrack->votes--;
+      }
+      if (pNormalTrack != nullptr) {
+        pNormalTrack->votes--;
+      }
+    }
+  } else {
+    // Track not in vote vector
+    if (vote) {
+      // Track not in vote vector and we want to upvote it: add to vector and
+      // update counter
+      it->votes.emplace_back(tID);
+      // increment its upvote counter
+      if (pAdminTrack != nullptr) {
+        pAdminTrack->votes++;
+      }
+      if (pNormalTrack != nullptr) {
+        pNormalTrack->votes++;
+      }
+    } else {
+      // track not in vote vector and we want to remove upvote: cant remove
+      // nonexistent upvote, so do nothing
+    }
+  }
+
+  // sort Normal Queue
+  sort(mNormalQueue.tracks.begin(), mNormalQueue.tracks.end());
+  return nullopt;
+}
+
+TResult<Queue> RAMDataStore::getQueue(QueueType q) {
+  // Shared Access to Song Queue
+  shared_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  // select Queue
+  Queue *pQueue = SelectQueue(q);
+  if (pQueue == nullptr) {
+    return Error(ErrorCode::InvalidValue, "Invalid Parameter in Queue");
+  }
+
+  // return read only access to Queue
+  return (const Queue)(*pQueue);
+}
+
+TResult<PlaybackTrack> RAMDataStore::getPlayingTrack() {
+  // Shared Access to Song Queue
+  shared_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  return mCurrentTrack;
+}
+
+TResult<bool> RAMDataStore::hasUser(TSessionID ID) {
+  // Shared Access to User List
+  shared_lock<shared_mutex> MyLock(mUserMutex, defer_lock);
+  MyLock.lock();
+
+  // find user
+  User user;
+  user.SessionID = ID;
+  auto it = find(mUsers.begin(), mUsers.end(), user);
+  if (it == mUsers.end()) {
+    return false;
+  } else {
+    return true;
+  }
+}
+
+TResultOpt RAMDataStore::nextTrack() {
+  // Exclusive Access to Song Queue
+  unique_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  // Increment LastPlayed counter for all songs
+
+  // doesnt work with auto for some reason
+  //  for(auto elem : mAdminQueue.tracks){
```

Comment:
that copies `elem` out of the container. of course incrementing does not effect the element in the container

Reply:
well that explains it. thanks